src/Main.hs

```diff
@@ -90,9 +90,22 @@ getExecutables = do
         [] -> error "No executables found"
         ds -> map fst ds
 
+getCabalProjectRootCurrent :: IO (Maybe FilePath)
+getCabalProjectRootCurrent = flip catchIOError (const (return Nothing)) $ do
+  cwd <- getCurrentDirectory
+  es  <- getDirectoryContents cwd
+  if any (isSuffixOf ".cabal") es
+    then return $ Just cwd
+    else do
+      let parent = takeDirectory cwd
+      if parent == cwd
+        then return Nothing
+        else withCurrentDirectory parent getCabalProjectRootCurrent
```

Comment:
I didn't realise `withCurrentDirectory` was a recent addition. I will reimplement it as you described.